### PR TITLE
Allow adding/removing of Datasets by direct reference only

### PR DESCRIPTION
**Pull request by jtlan**
Users have to pass in a `Dataset` to add or remove it. No more arrays or keys!

The internal implementation still uses auto-generated keys, but we can patch that up later.

### CODE REVIEW

File: src/components/plots/plot.ts [hunk 1]
```diff
@@ -514,27 +486,16 @@ module Plottable {
...
-        datasetKeyArray = [<string> datasetKeys];
-      } else {
-        datasetKeyArray = <string[]> datasetKeys;
-      }
-
-      return this._getAllPlotData(datasetKeyArray);
-    }
-
-    protected _getAllPlotData(datasetKeys: string[]): Plots.PlotData {
+    public getAllPlotData(datasets = this.datasets()): Plots.PlotData {
```
Comment (bluong) [inline src/components/plots/plot.ts]: Why can't we allow for Dataset | Dataset[]?  Our architecture isn't hurt that much and it allows for a nice streamlined API

File: src/components/plots/plot.ts [hunk 2]
```diff
@@ -83,39 +83,23 @@ module Plottable {
...
-        throw new Error("invalid input to addDataset");
-      }
-      if (typeof(keyOrDataset) === "string" && keyOrDataset[0] === "_") {
-        Utils.Methods.warn("Warning: Using _named series keys may produce collisions with unlabeled data sources");
-      }
-      var key  = typeof(keyOrDataset) === "string" ? keyOrDataset : "_" + this._nextSeriesIndex++;
-      var data = typeof(keyOrDataset) !== "string" ? keyOrDataset : dataset;
-      dataset = (data instanceof Dataset) ? data : new Dataset(data);
-
+    public addDataset(dataset: Dataset) {
```
Comment (bluong) [inline src/components/plots/plot.ts]: Do you think the doc is enough to convey what this does?

File: src/components/plots/plot.ts [hunk 3]
```diff
@@ -83,39 +83,23 @@ module Plottable {
...
+      var key = "_" + this._nextSeriesIndex++;
```
Comment (bluong) [inline src/components/plots/plot.ts]: On a side note, if dataset's had their own identifiers, this is resolved

File: src/components/plots/plot.ts [hunk 4]
```diff
@@ -224,18 +208,21 @@ module Plottable {
...
-      var userMetadata = plotDatasetKey.dataset.metadata();
+    public generateProjectors(dataset: Dataset): AttributeToAppliedProjector {
       var attrToAppliedProjector: AttributeToAppliedProjector = {};
-      d3.entries(attrToProjector).forEach((keyValue: any) => {
-        attrToAppliedProjector[keyValue.key] = (datum: any, index: number) => keyValue.value(datum, index, userMetadata, plotMetadata);
-      });
+      var datasetKey = this._keyForDataset(dataset);
+      if (datasetKey != null) {
+        var attrToProjector = this._generateAttrToProjector();
+        var plotDatasetKey = this._key2PlotDatasetKey.get(datasetKey);
```
Comment (bluong) [inline src/components/plots/plot.ts]: Doesn't really matter for the user / it's only an architectural complaint, but we really can just have a _dataset2PlotDatasetKey (or make that even cleaner)

File: test/components/plots/plotTests.ts [hunk 5]
```diff
@@ -151,35 +142,37 @@ describe("Plots", () => {
 
       var renderArea2 = svg.append("g");
       renderArea2.append("circle").attr("cx", 10).attr("cy", 10).attr("r", 10);
-      var mockDrawer2 = new Plottable.Drawers.AbstractDrawer("ds2");
+      var mockDrawer2 = new Plottable.Drawers.AbstractDrawer("_1"); // HACKHACK: Dataset keys are being removed, so this is the internal key
```
Comment (bluong) [inline test/components/plots/plotTests.ts]: We can move such that Drawers don't use string keys as well

File: test/components/plots/stackedAreaPlotTests.ts [hunk 6]
```diff
@@ -393,36 +399,39 @@ describe("Plots", () => {
...
+      plot.addDataset(dataset2);
       plot.project("fill", "fill");
       plot.project("x", "x", xScale).project("y", "y", yScale);
 
-      var ds1Point2Offset = (<any> plot)._key2PlotDatasetKey.get("d1").plotMetadata.offsets.get(2);
-      var ds2Point2Offset = (<any> plot)._key2PlotDatasetKey.get("d2").plotMetadata.offsets.get(2);
-      var ds3Point2Offset = (<any> plot)._key2PlotDatasetKey.get("d3").plotMetadata.offsets.get(2);
+      var ds0Point2Offset = (<any> plot)._key2PlotDatasetKey.get("_0").plotMetadata.offsets.get(2);
+      var ds1Point2Offset = (<any> plot)._key2PlotDatasetKey.get("_1").plotMetadata.offsets.get(2);
+      var ds2Point2Offset = (<any> plot)._key2PlotDatasetKey.get("_2").plotMetadata.offsets.get(2);
```
Comment (bluong) [inline test/components/plots/stackedAreaPlotTests.ts]: Let's put a better note that we need to remove plotMetadata

File: test/components/plots/stackedBarPlotTests.ts [hunk 7]
```diff
@@ -418,19 +418,19 @@ describe("Plots", () => {
...
-      plot.addDataset("d2", data2);
+      plot.addDataset(new Plottable.Dataset(data1));
+      plot.addDataset(new Plottable.Dataset(data2));
       plot.project("fill", "fill");
       plot.project("x", "x", xScale).project("y", "y", yScale);
 
-      var ds1FirstColumnOffset = (<any> plot)._key2PlotDatasetKey.get("d1").plotMetadata.offsets.get("A");
-      var ds2FirstColumnOffset = (<any> plot)._key2PlotDatasetKey.get("d2").plotMetadata.offsets.get("A");
+      var ds1FirstColumnOffset = (<any> plot)._key2PlotDatasetKey.get("_0").plotMetadata.offsets.get("A");
+      var ds2FirstColumnOffset = (<any> plot)._key2PlotDatasetKey.get("_1").plotMetadata.offsets.get("A");
```
Comment (bluong) [inline test/components/plots/stackedBarPlotTests.ts]: Same as above

File: test/components/plots/stackedPlotTests.ts [hunk 8]
```diff
@@ -19,145 +19,145 @@ describe("Plots", () => {
...
+      stackedPlot.addDataset(new Plottable.Dataset(data1));
+      stackedPlot.addDataset(new Plottable.Dataset(data2));
+      stackedPlot.addDataset(new Plottable.Dataset(data3));
+      stackedPlot.addDataset(new Plottable.Dataset(data4));
+
+      var keys = (<any> stackedPlot)._key2PlotDatasetKey.keys(); // HACKHACK: Dataset keys are being removed, so these are internal keys
+      var ds1PlotMetadata = <Plottable.Plots.StackedPlotMetadata>(<any> stackedPlot)._key2PlotDatasetKey.get(keys[1]).plotMetadata;
+      var ds4PlotMetadata = <Plottable.Plots.StackedPlotMetadata>(<any> stackedPlot)._key2PlotDatasetKey.get(keys[4]).plotMetadata;
+      assert.strictEqual(ds1PlotMetadata.offsets.get("1"), 1, "positive offset was used");
+      assert.strictEqual(ds4PlotMetadata.offsets.get("1"), 2, "positive offset was used");
```
Comment (bluong) [inline test/components/plots/stackedPlotTests.ts]: Same as above

File: test/components/plots/stackedPlotTests.ts [hunk 9]
```diff
@@ -19,145 +19,145 @@ describe("Plots", () => {
...
     });
-
+//
```
Comment (bluong) [inline test/components/plots/stackedPlotTests.ts]: ?? Remove comment

File: test/components/plots/stackedPlotTests.ts [hunk 10]
```diff
@@ -19,145 +19,145 @@ describe("Plots", () => {
...
+      stackedPlot.addDataset(new Plottable.Dataset(data2));
+      stackedPlot.addDataset(new Plottable.Dataset(data3));
+      stackedPlot.addDataset(new Plottable.Dataset(data4));
+      stackedPlot.addDataset(new Plottable.Dataset(data5));
 
-      var ds3PlotMetadata = <Plottable.Plots.StackedPlotMetadata>(<any> stackedPlot)._key2PlotDatasetKey.get("d3").plotMetadata;
-      var ds4PlotMetadata = <Plottable.Plots.StackedPlotMetadata>(<any> stackedPlot)._key2PlotDatasetKey.get("d4").plotMetadata;
-      var ds5PlotMetadata = <Plottable.Plots.StackedPlotMetadata>(<any> stackedPlot)._key2PlotDatasetKey.get("d5").plotMetadata;
-      var ds6PlotMetadata = <Plottable.Plots.StackedPlotMetadata>(<any> stackedPlot)._key2PlotDatasetKey.get("d6").plotMetadata;
+      var keys = (<any> stackedPlot)._key2PlotDatasetKey.keys(); // HACKHACK: Dataset keys are being removed, so these are internal keys
```
Comment (bluong) [inline test/components/plots/stackedPlotTests.ts]: This is mostly a plotMetadata issue but we should probably ticket this for reference

File: test/drawers/arcDrawerTests.ts [hunk 11]
```diff
@@ -7,10 +7,10 @@ describe("Drawers", () => {
       var data = [{value: 10}, {value: 10}, {value: 10}, {value: 10}];
       var piePlot = new Plottable.Plots.Pie();
 
-      var drawer = new Plottable.Drawers.Arc("one");
+      var drawer = new Plottable.Drawers.Arc("_0"); // HACKHACK: Dataset keys are being removed, so this is the internal key
```
Comment (bluong) [inline test/drawers/arcDrawerTests.ts]: Let's track this?

File: test/scales/scaleTests.ts [hunk 12]
```diff
@@ -154,7 +154,7 @@ describe("Scales", () => {
     });
   });
 
-  describe("QuantitativeScale Scales", () => {
+  describe("Quantitative Scales", () => {
```
Comment (bluong) [inline test/scales/scaleTests.ts]: The previous name tho...

**Comment by bluong** [comment]
I guess the biggest thing I have to say is why can't we all getAllPlotData(dataset)

File: src/components/plots/plot.ts (hunk 2 above)
Comment (jtlan) [inline src/components/plots/plot.ts]: It's called `addDataset()`, so I think it's safe to assume that it adds a `Dataset`.

File: src/components/plots/plot.ts (hunk 3 above)
Comment (jtlan) [inline src/components/plots/plot.ts]: Should they have IDs? It's possible that it makes sense for `Dataset`s, but nothing else.

The other way to do this is to use a `Map` of some sort, but that's a bigger touch refactor.

File: src/components/plots/plot.ts (hunk 4 above)
Comment (jtlan) [inline src/components/plots/plot.ts]: That's be a `Map` of some sort, right? We can do that refactor later, the main point was to get the public-facing API in a good spot.

File: test/components/plots/plotTests.ts (hunk 5 above)
Comment (jtlan) [inline test/components/plots/plotTests.ts]: And just take the `Dataset`?

**Comment by jtlan** [comment]
@bluong wrote:
>I guess the biggest thing I have to say is why can't we all getAllPlotData(dataset)

I guess it's reasonably discoverable because of the typing info.

File: test/components/plots/plotTests.ts (hunk 5 above)
Comment (bluong) [inline test/components/plots/plotTests.ts]: Yup

File: src/components/plots/plot.ts (hunk 3 above)
Comment (bluong) [inline src/components/plots/plot.ts]: I think so.  Out of scope, but yes I think they should have ids

**Pull request merged**